File: exporter.py
"""
exporter.py
-----------
Writes completed scan results out to disk in TXT, CSV, or JSON format.
Each export includes the target, scan date, duration, the full port
list, service names, and summary statistics.
"""

from __future__ import annotations

import csv
import json
from dataclasses import asdict
from pathlib import Path
from typing import Iterable

from scanner import ScanResult
from utils import ensure_dir, now_stamp_for_filename


def _summary(target: str, results: Iterable[ScanResult], duration: float) -> dict:
    results = list(results)
    open_ports = [r for r in results if r.status == "open"]
    closed_ports = [r for r in results if r.status == "closed"]
    filtered_ports = [r for r in results if r.status not in ("open", "closed")]
    return {
        "target": target,
        "scan_date": results[0].scanned_at if results else "",
        "duration_seconds": round(duration, 2),
        "total_ports": len(results),
        "open_count": len(open_ports),
        "closed_count": len(closed_ports),
        "filtered_count": len(filtered_ports),
    }


def build_export_path(export_folder: str, target: str, fmt: str) -> Path:
    """Build a timestamped output path inside the configured export folder."""
    folder = ensure_dir(export_folder)
    safe_target = target.replace(":", "_").replace("/", "_")
    filename = f"ndportx_{safe_target}_{now_stamp_for_filename()}.{fmt}"
    return folder / filename


def export_txt(path: Path, target: str, results: list[ScanResult], duration: float) -> None:
    summary = _summary(target, results, duration)
    lines = [
        "NDPortX - Scan Report",
        "=" * 40,
        f"Target:        {summary['target']}",
        f"Scan date:     {summary['scan_date']}",
        f"Duration:      {summary['duration_seconds']}s",
        f"Total ports:   {summary['total_ports']}",
        f"Open ports:    {summary['open_count']}",
        f"Closed ports:  {summary['closed_count']}",
        f"Filtered:      {summary['filtered_count']}",
        "-" * 40,
        "PORT     STATUS     SERVICE          BANNER",
    ]
    for r in results:
        banner = (r.banner or "")[:40]
        lines.append(f"{r.port:<8} {r.status:<10} {r.service:<16} {banner}")
    path.write_text("\n".join(lines), encoding="utf-8")


def export_csv(path: Path, target: str, results: list[ScanResult], duration: float) -> None:
    summary = _summary(target, results, duration)
    with path.open("w", newline="", encoding="utf-8") as f:
        writer = csv.writer(f)
        writer.writerow(["# Target", summary["target"]])
        writer.writerow(["# Scan date", summary["scan_date"]])
        writer.writerow(["# Duration (s)", summary["duration_seconds"]])
        writer.writerow(["# Open", summary["open_count"]])
        writer.writerow(["# Closed", summary["closed_count"]])
        writer.writerow(["# Filtered", summary["filtered_count"]])
        writer.writerow([])
        writer.writerow(["Port", "Status", "Service", "Banner", "Response Time (ms)"])
        for r in results:
            writer.writerow([r.port, r.status, r.service, r.banner or "", r.response_ms])


def export_json(path: Path, target: str, results: list[ScanResult], duration: float) -> None:
    summary = _summary(target, results, duration)
    data = {
        "summary": summary,
        "results": [asdict(r) for r in results],
    }
    path.write_text(json.dumps(data, indent=2), encoding="utf-8")


EXPORTERS = {
    "txt": export_txt,
    "csv": export_csv,
    "json": export_json,
}


def export_results(
    export_folder: str,
    target: str,
    results: list[ScanResult],
    duration: float,
    fmt: str,
) -> Path:
    """Export results in the requested format and return the file path."""
    fmt = fmt.lower().lstrip(".")
    if fmt not in EXPORTERS:
        raise ValueError(f"Unsupported export format: {fmt}")
    path = build_export_path(export_folder, target, fmt)
    EXPORTERS[fmt](path, target, results, duration)
    return path

File: src/exporter.ts
import fs from "node:fs";
import path from "node:path";
import { ScanResult } from "./scanner.js";
import { ensureDir, nowStampForFilename } from "./utils.js";

export interface ScanSummary {
  target: string;
  scan_date: string;
  duration_seconds: number;
  total_ports: number;
  open_count: number;
  closed_count: number;
  filtered_count: number;
}

export function buildSummary(target: string, results: ScanResult[], duration: number): ScanSummary {
  const open = results.filter((r) => r.status === "open");
  const closed = results.filter((r) => r.status === "closed");
  const filtered = results.filter((r) => r.status !== "open" && r.status !== "closed");
  return {
    target,
    scan_date: results[0]?.scanned_at || "",
    duration_seconds: Math.round(duration * 100) / 100,
    total_ports: results.length,
    open_count: open.length,
    closed_count: closed.length,
    filtered_count: filtered.length,
  };
}

export function buildExportFilename(target: string, fmt: string): string {
  const safeTarget = target.replace(/[:/\\]/g, "_");
  return `ndportx_${safeTarget}_${nowStampForFilename()}.${fmt}`;
}

export function exportTxt(target: string, results: ScanResult[], duration: number): string {
  const summary = buildSummary(target, results, duration);
  const lines = [
    "NDPortX - Scan Report",
    "=".repeat(40),
    `Target:        ${summary.target}`,
    `Scan date:     ${summary.scan_date}`,
    `Duration:      ${summary.duration_seconds}s`,
    `Total ports:   ${summary.total_ports}`,
    `Open ports:    ${summary.open_count}`,
    `Closed ports:  ${summary.closed_count}`,
    `Filtered:      ${summary.filtered_count}`,
    "-".repeat(40),
    "PORT     STATUS     SERVICE          BANNER",
  ];
  for (const r of results) {
    const banner = (r.banner || "").slice(0, 40);
    const p = String(r.port).padEnd(8);
    const st = r.status.padEnd(10);
    const sv = r.service.padEnd(16);
    lines.push(`${p} ${st} ${sv} ${banner}`);
  }
  return lines.join("\n");
}

export function exportCsv(target: string, results: ScanResult[], duration: number): string {
  const summary = buildSummary(target, results, duration);
  const lines = [
    `# Target,${summary.target}`,
    `# Scan date,${summary.scan_date}`,
    `# Duration (s),${summary.duration_seconds}`,
    `# Open,${summary.open_count}`,
    `# Closed,${summary.closed_count}`,
    `# Filtered,${summary.filtered_count}`,
    "",
    "Port,Status,Service,Banner,Response Time (ms)",
  ];
  for (const r of results) {
    const banner = (r.banner || "").replace(/"/g, '""');
    lines.push(`${r.port},${r.status},${r.service},"${banner}",${r.response_ms}`);
  }
  return lines.join("\n");
}

export function exportJson(target: string, results: ScanResult[], duration: number): string {
  const summary = buildSummary(target, results, duration);
  return JSON.stringify(
    {
      summary,
      results,
    },
    null,
    2
  );
}

export function exportResults(
  exportFolder: string,
  target: string,
  results: ScanResult[],
  duration: number,
  fmt: string
): { filePath: string; filename: string; content: string; mimeType: string } {
  const cleanFmt = fmt.toLowerCase().replace(/^\./, "");
  ensureDir(exportFolder);
  const filename = buildExportFilename(target, cleanFmt);
  const filePath = path.join(exportFolder, filename);

  let content = "";
  let mimeType = "text/plain";

  if (cleanFmt === "csv") {
    content = exportCsv(target, results, duration);
    mimeType = "text/csv";
  } else if (cleanFmt === "json") {
    content = exportJson(target, results, duration);
    mimeType = "application/json";
  } else if (cleanFmt === "txt") {
    content = exportTxt(target, results, duration);
    mimeType = "text/plain";
  } else {
    throw new Error(`Unsupported export format: ${fmt}`);
  }

  fs.writeFileSync(filePath, content, "utf-8");
  return { filePath, filename, content, mimeType };
}
